### ISSUE TITLE
IstioOperator using HorizontalPodAutoscalerSpec v1alpha1

### ISSUE THREAD

**Issue by SantiMunoz** [istio/istio#39448]
### Bug Description

The [IstioOperator documentation](https://istio.io/latest/docs/reference/config/istio.operator.v1alpha1/#HPAScalingRules) states that we can use the `behavior` field of the HPA for our istiod deployment, but when applying the below configuration we get the error `Failed to watch *v1alpha1.IstioOperator: failed to list *v1alpha1.IstioOperator: unknown field "behavior" in v1alpha1.HorizontalPodAutoscalerSpec`

```yaml
components:
    pilot:
      k8s:
        hpaSpec:
          maxReplicas: 10
          minReplicas: 3
          behavior:
            scaleDown:
              stabilizationWindowSeconds: 3600
```

### Version

```prose
➜  ~ istioctl version
client version: 1.12.2
control plane version: 1.13.2
➜  ~ kubectl version --short
Client Version: v1.20.1
Server Version: v1.22.10
```

**Comment by litong01** [comment]
@howardjohn @SantiMunoz I've tried 1.13.3, the issue still exists, however, it worked fine on 1.14.1. I think that the option was added maybe after 1.13.x.

**Comment by litong01** [comment]
@howardjohn I wonder if somehow we can indicate when an option being made available. I think some of the products show options with `since` tag which can give people clues on if an option should be working for the release they are using.

**Comment by KunalMalhotraN26** [comment]
@howardjohn Yup just tested with 1.14.1, no error anymore.

**Comment by SantiMunoz** [comment]
Thanks everyone, I've just tested it with 1.14.1 and the error is gone, I'm just a little bit confused with the HPA generated by the operator, it has the following definition:

```yaml
apiVersion: autoscaling/v2beta1
kind: HorizontalPodAutoscaler
metadata:
  name: istiod
  namespace: istio-system
  uid: f6398566-2340-4bdc-a1b7-b3e6d3080911
  resourceVersion: '59128557'
  creationTimestamp: '2022-06-15T07:46:44Z'
  labels:
    app: istiod
    install.operator.istio.io/owning-resource: istio-controlplane
    install.operator.istio.io/owning-resource-namespace: istio-system
    istio.io/rev: default
    operator.istio.io/component: Pilot
    operator.istio.io/managed: Reconcile
    operator.istio.io/version: 1.14.1
    release: istio
  annotations:
    autoscaling.alpha.kubernetes.io/behavior: >-
      {"ScaleUp":{"StabilizationWindowSeconds":0,"SelectPolicy":"Max","Policies":[{"Type":"Pods","Value":4,"PeriodSeconds":15},{"Type":"Percent","Value":100,"PeriodSeconds":15}]},"ScaleDown":{"StabilizationWindowSeconds":3600,"SelectPolicy":"Max","Policies":[{"Type":"Percent","Value":100,"PeriodSeconds":15}]}}
    kubectl.kubernetes.io/last-applied-configuration: >
      {"apiVersion":"autoscaling/v2beta2","kind":"HorizontalPodAutoscaler","metadata":{"annotations":{},"labels":{"app":"istiod","install.operator.istio.io/owning-resource":"istio-controlplane","install.operator.istio.io/owning-resource-namespace":"istio-system","istio.io/rev":"default","operator.istio.io/component":"Pilot","operator.istio.io/managed":"Reconcile","operator.istio.io/version":"1.14.1","release":"istio"},"name":"istiod","namespace":"istio-system"},"spec":{"behavior":{"scaleDown":{"stabilizationWindowSeconds":3600}},"maxReplicas":10,"metrics":[{"resource":{"name":"cpu","target":{"averageUtilization":80,"type":"Utilization"}},"type":"Resource"}],"minReplicas":3,"scaleTargetRef":{"apiVersion":"apps/v1","kind":"Deployment","name":"istiod"}}}
...<REDACTED>...
spec:
  scaleTargetRef:
    kind: Deployment
    name: istiod
    apiVersion: apps/v1
  minReplicas: 3
  maxReplicas: 10
  metrics:
    - type: Resource
      resource:
        name: cpu
        targetAverageUtilization: 80
```

In the annotations it mentions that it has applied the `stabilizationWindowSeconds`, but in the `spec` of the HPA it doesn't have the `behaviour` part.
Another mismatch is the api version, the HPA is using v2beta1 while the `kubectl.kubernetes.io/last-applied-configuration` annotation says `v2beta2`.

Could you help me understand this configuration drift between the `kubectl.kubernetes.io/last-applied-configuration` and trhe `spec`?

**Comment by litong01** [comment]
@SantiMunoz can you post the complete yaml file here?

**Comment by SantiMunoz** [comment]
Hey @litong01, thanks for the follow up, find below the HPA I see for istiod:

```yaml
apiVersion: autoscaling/v2beta1
kind: HorizontalPodAutoscaler
metadata:
  name: istiod
  namespace: istio-system
  uid: f6398566-2340-4bdc-a1b7-b3e6d3080911
  resourceVersion: '59569661'
  creationTimestamp: '2022-06-15T07:46:44Z'
  labels:
    app: istiod
    install.operator.istio.io/owning-resource: istio-controlplane
    install.operator.istio.io/owning-resource-namespace: istio-system
    istio.io/rev: default
    operator.istio.io/component: Pilot
    operator.istio.io/managed: Reconcile
    operator.istio.io/version: 1.14.1
    release: istio
  annotations:
    autoscaling.alpha.kubernetes.io/behavior: >-
      {"ScaleUp":{"StabilizationWindowSeconds":0,"SelectPolicy":"Max","Policies":[{"Type":"Pods","Value":4,"PeriodSeconds":15},{"Type":"Percent","Value":100,"PeriodSeconds":15}]},"ScaleDown":{"StabilizationWindowSeconds":3600,"SelectPolicy":"Max","Policies":[{"Type":"Percent","Value":100,"PeriodSeconds":15}]}}
    kubectl.kubernetes.io/last-applied-configuration: >
      {"apiVersion":"autoscaling/v2beta2","kind":"HorizontalPodAutoscaler","metadata":{"annotations":{},"labels":{"app":"istiod","install.operator.istio.io/owning-resource":"istio-controlplane","install.operator.istio.io/owning-resource-namespace":"istio-system","istio.io/rev":"default","operator.istio.io/component":"Pilot","operator.istio.io/managed":"Reconcile","operator.istio.io/version":"1.14.1","release":"istio"},"name":"istiod","namespace":"istio-system"},"spec":{"behavior":{"scaleDown":{"stabilizationWindowSeconds":3600}},"maxReplicas":10,"metrics":[{"resource":{"name":"cpu","target":{"averageUtilization":80,"type":"Utilization"}},"type":"Resource"}],"minReplicas":3,"scaleTargetRef":{"apiVersion":"apps/v1","kind":"Deployment","name":"istiod"}}}
  managedFields:
    - manager: istio-operator
      operation: Apply
      apiVersion: autoscaling/v2beta2
      time: '2022-06-15T07:46:44Z'
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:annotations:
            f:kubectl.kubernetes.io/last-applied-configuration: {}
          f:labels:
            f:app: {}
            f:install.operator.istio.io/owning-resource: {}
            f:install.operator.istio.io/owning-resource-namespace: {}
            f:istio.io/rev: {}
            f:operator.istio.io/component: {}
            f:operator.istio.io/managed: {}
            f:operator.istio.io/version: {}
            f:release: {}
        f:spec:
          f:behavior:
            f:scaleDown:
              f:stabilizationWindowSeconds: {}
          f:maxReplicas: {}
          f:metrics: {}
          f:minReplicas: {}
          f:scaleTargetRef:
            f:apiVersion: {}
            f:kind: {}
            f:name: {}
    - manager: kube-controller-manager
      operation: Update
      apiVersion: autoscaling/v1
      time: '2022-06-15T07:47:16Z'
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:annotations:
            f:autoscaling.alpha.kubernetes.io/conditions: {}
            f:autoscaling.alpha.kubernetes.io/current-metrics: {}
        f:status:
          f:currentCPUUtilizationPercentage: {}
          f:currentReplicas: {}
          f:desiredReplicas: {}
  selfLink: >-
    /apis/autoscaling/v2beta1/namespaces/istio-system/horizontalpodautoscalers/istiod
status:
  currentReplicas: 3
  desiredReplicas: 3
  currentMetrics:
    - type: Resource
      resource:
        name: cpu
        currentAverageUtilization: 1
        currentAverageValue: 5m
  conditions:
    - type: AbleToScale
      status: 'True'
      lastTransitionTime: '2022-06-15T07:47:16Z'
      reason: ReadyForNewScale
      message: recommended size matches current size
    - type: ScalingActive
      status: 'True'
      lastTransitionTime: '2022-06-15T07:47:16Z'
      reason: ValidMetricFound
      message: >-
        the HPA was able to successfully calculate a replica count from cpu
        resource utilization (percentage of request)
    - type: ScalingLimited
      status: 'True'
      lastTransitionTime: '2022-06-15T08:47:41Z'
      reason: TooFewReplicas
      message: the desired replica count is less than the minimum replica count
spec:
  scaleTargetRef:
    kind: Deployment
    name: istiod
    apiVersion: apps/v1
  minReplicas: 3
  maxReplicas: 10
  metrics:
    - type: Resource
      resource:
        name: cpu
        targetAverageUtilization: 80

```

**Comment by litong01** [comment]
@SantiMunoz Here is the yaml file I used.

```
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  namespace: istio-system
  name: example-istiocontrolplane
spec:
  profile: default
  components:
    pilot:
      k8s:
        hpaSpec:
          maxReplicas: 10
          minReplicas: 3
          behavior:
            scaleDown:
              stabilizationWindowSeconds: 3600
```

And once the PodAutoscaler is up, I am seeing the following.

```
apiVersion: autoscaling/v2
kind: HorizontalPodAutoscaler
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"autoscaling/v2beta2","kind":"HorizontalPodAutoscaler","metadata":{"annotations":{},"labels":{"app":"istiod","install.operator.istio.io/owning-resource":"example-istiocontrolplane","install.operator.istio.io/owning-resource-namespace":"istio-system","istio.io/rev":"default","operator.istio.io/component":"Pilot","operator.istio.io/managed":"Reconcile","operator.istio.io/version":"1.14.1","release":"istio"},"name":"istiod","namespace":"istio-system"},"spec":{"behavior":{"scaleDown":{"stabilizationWindowSeconds":3600}},"maxReplicas":10,"metrics":[{"resource":{"name":"cpu","target":{"averageUtilization":80,"type":"Utilization"}},"type":"Resource"}],"minReplicas":3,"scaleTargetRef":{"apiVersion":"apps/v1","kind":"Deployment","name":"istiod"}}}
  creationTimestamp: "2022-06-15T14:18:22Z"
  labels:
    app: istiod
    install.operator.istio.io/owning-resource: example-istiocontrolplane
    install.operator.istio.io/owning-resource-namespace: istio-system
    istio.io/rev: default
    operator.istio.io/component: Pilot
    operator.istio.io/managed: Reconcile
    operator.istio.io/version: 1.14.1
    release: istio
  managedFields:
  - apiVersion: autoscaling/v2beta2
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          f:kubectl.kubernetes.io/last-applied-configuration: {}
        f:labels:
          f:app: {}
          f:install.operator.istio.io/owning-resource: {}
          f:install.operator.istio.io/owning-resource-namespace: {}
          f:istio.io/rev: {}
          f:operator.istio.io/component: {}
          f:operator.istio.io/managed: {}
          f:operator.istio.io/version: {}
          f:release: {}
      f:spec:
        f:behavior:
          f:scaleDown:
            f:stabilizationWindowSeconds: {}
        f:maxReplicas: {}
        f:metrics: {}
        f:minReplicas: {}
        f:scaleTargetRef:
          f:apiVersion: {}
          f:kind: {}
          f:name: {}
    manager: istio-operator
    operation: Apply
    time: "2022-06-15T14:18:22Z"
  - apiVersion: autoscaling/v2
    fieldsType: FieldsV1
    fieldsV1:
      f:status:
        f:conditions:
          .: {}
          k:{"type":"AbleToScale"}:
            .: {}
            f:lastTransitionTime: {}
            f:message: {}
            f:reason: {}
            f:status: {}
            f:type: {}
          k:{"type":"ScalingActive"}:
            .: {}
            f:lastTransitionTime: {}
            f:message: {}
            f:reason: {}
            f:status: {}
            f:type: {}
        f:currentReplicas: {}
        f:desiredReplicas: {}
        f:lastScaleTime: {}
    manager: kube-controller-manager
    operation: Update
    subresource: status
    time: "2022-06-15T14:18:52Z"
  name: istiod
  namespace: istio-system
  resourceVersion: "3810"
  uid: d88bcf17-c8ff-477d-9263-0d96a843e10d
spec:
  behavior:
    scaleDown:
      policies:
      - periodSeconds: 15
        type: Percent
        value: 100
      selectPolicy: Max
      stabilizationWindowSeconds: 3600
    scaleUp:
      policies:
      - periodSeconds: 15
        type: Pods
        value: 4
      - periodSeconds: 15
        type: Percent
        value: 100
      selectPolicy: Max
      stabilizationWindowSeconds: 0
  maxReplicas: 10
  metrics:
  - resource:
      name: cpu
      target:
        averageUtilization: 80
        type: Utilization
    type: Resource
  minReplicas: 3
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: istiod
status:
  conditions:
  - lastTransitionTime: "2022-06-15T14:18:37Z"
    message: the HPA controller was able to get the target's current scale
    reason: SucceededGetScale
    status: "True"
    type: AbleToScale
  - lastTransitionTime: "2022-06-15T14:18:52Z"
    message: 'the HPA was unable to compute the replica count: failed to get cpu utilization:
      unable to get metrics for resource cpu: unable to fetch metrics from resource
      metrics API: the server could not find the requested resource (get pods.metrics.k8s.io)'
    reason: FailedGetResourceMetric
    status: "False"
    type: ScalingActive
  currentMetrics: null
  currentReplicas: 3
  desiredReplicas: 3
  lastScaleTime: "2022-06-15T14:18:37Z"
```

notice that the api version is `apiVersion: autoscaling/v2` and behavior section also contains the correct  `stabilizationWindowSeconds`

**Comment by litong01** [comment]
@SantiMunoz my env is clean. The version issue that you saw maybe due to some residual of older istio resources?

**Comment by SantiMunoz** [comment]
@litong01 We are using Kubernetes 1.22, I think that is the reason we see different object versions. The Istio Operator we use is very similar the one you shared so I don't think the issue is coming from there.

I will try to test it in k8s 1.23 if I have the chance :)

**Comment by SantiMunoz** [comment]
I will close the issue as it seems to be fixed in k8s 1.23, but I would like to emphasise the idea of having the minimum requirements per feature in the Istio docs, that would help a lot! 

Thanks everyone for the help!
